concurrency/TransportBuddyManager.java

```diff
@@ -0,0 +1,270 @@
+                B buddy = getBuddy(jid);
+                buddy.setVerboseStatus(pendingVerboseStatuses.get(jid));
+            }
+            catch (NotFoundException e) {
+                // Alrighty then....
+            }
+        }
+        pendingPresences.clear();
+        pendingVerboseStatuses.clear();
+        isActive = true;
+        sendAllAvailablePresences(getSession().getJID());
+    }
+
+    /**
+     * Stores a status setting to be recorded after the buddy manager has been activated.
+     * This is typically used when status's arrive before the buddy list is activated.
+     *
+     * @param jid JID of the contact to set status for later.
+     * @param presence Presence of contact.
+     * @param status verbose status string of contact.
+     */
+    public synchronized void storePendingStatus(JID jid, PresenceType presence, String status) {
+        if (!isActivated()) {
+            pendingPresences.put(jid, presence);
+            pendingVerboseStatuses.put(jid, status);
+        }
+        else {
+            try {
+                B buddy = getBuddy(jid);
+                buddy.setPresenceAndStatus(presence, status);
+            }
+            catch (NotFoundException e) {
+                // Alrighty then....
+            }
+        }
+    }
+
+    /**
+     * Resets (clears) the list of buddies assigned to the buddy manager.
+     */
+    public void resetBuddies() {
+        buddies.clear();
+    }
+
+    /**
+     * Retrieve the buddy instance for a given JID.
+     *
+     * @param jid JID of the buddy to be retrieved.
+     * @throws NotFoundException if the given jid is not found.
+     * @return TransportBuddy instance requested.
+     */
+    public B getBuddy(JID jid) throws NotFoundException {
+        B buddy = buddies.get(jid);
+        if (buddy == null) {
+            throw new NotFoundException("Could not find buddy requested.");
+        }
+        return buddy;
+    }
+    
+    /**
+     * Retrieve the buddy instance for a given user.
+     *
+     * @param username Username of the buddy to be retrieved.
+     * @throws NotFoundException if the given username is not found.
+     * @return TransportBuddy instance requested.
+     */
+// Disabling this because it can cause confusion in one's attempt to match a buddy.
+//    public B getBuddy(String username) throws NotFoundException {
+//        B buddy = buddies.get(username.toLowerCase());
+//        if (buddy == null) {
+//            throw new NotFoundException("Could not find buddy requested.");
+//        }
+//        return buddy;
+//    }
+
+    /**
+     * Stores a new buddy instance in the buddy list.
+     *
+     * @param buddy TransportBuddy associated with the username.
+     */
+    public void storeBuddy(B buddy) {
+        if (!buddies.containsKey(buddy.jid)) {
+            Log.debug("("+getSession().getTransport().getType().toString().toUpperCase()+") Storing new buddy: "+buddy);
+            buddies.put(buddy.jid, buddy);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **174 and 176**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and uses that reference at line: **180**, **185**, **178**, **180**, **180**, and **185**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.